### PR TITLE
Treap updates

### PR DESCRIPTION
**Pull request by dutinmeow**

### CODE REVIEW

**Comment by dutinmeow** [comment]
just ignore the stuff i did trying to create an asset folder

File: content/6_Advanced/Treaps.mdx [hunk 1]
```diff
@@ -1,27 +1,128 @@
...
+
+void merge(Node &*root, Node *left, Node *right) {
+	if (!l || !r) {
+		root = l ? l : r;
+		return;
+	}
+
+	if (left->pri > r->pri) {
+		merge(left->right, left->right, right);
+		root = left
```
Comment (bqi343) [inline content/6_Advanced/Treaps.mdx:114]: .

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: content/6_Advanced/Treaps.mdx (hunk 1 above)
Comment (dutinmeow) [reply content/6_Advanced/Treaps.mdx:114]: Wait what's wrong with this?

**Review by caoash — CHANGES_REQUESTED** [review]

File: content/6_Advanced/Treaps.mdx [hunk 2]
```diff
@@ -1,27 +1,128 @@
...
+	if (!root) {
+		left = right = NULL;
+		return;
+	}
+
+	if (root->val <= x) {
+		split(root->right, x, root->right, right);
+		left = root;
+	}
+	else {
```
Comment (SansPapyrus683) [inline content/6_Advanced/Treaps.mdx:100]:
```suggestion
	} else {
```

File: content/6_Advanced/Treaps.mdx [hunk 3]
```diff
@@ -1,27 +1,128 @@
...
+		split(root->left, x, left, root->left);
+		right = root;
+	}
+}
+
+void merge(Node &*root, Node *left, Node *right) {
+	if (!l || !r) {
```
Comment (SansPapyrus683) [inline content/6_Advanced/Treaps.mdx:107]: uuuuuuhhhh
where are `l` and `r` defined
i don't see them in the global scope
and they aren't in any of the arguments

File: content/6_Advanced/Treaps.mdx [hunk 4]
```diff
@@ -1,27 +1,128 @@
...
+		root = l ? l : r;
+		return;
+	}
+
+	if (left->pri > r->pri) {
+		merge(left->right, left->right, right);
+		root = left;
+	}
+	else {
```
Comment (SansPapyrus683) [inline content/6_Advanced/Treaps.mdx:116]:
```suggestion
	} else {
```

File: content/6_Advanced/Treaps.mdx (hunk 3 above)
Comment (SansPapyrus683) [reply content/6_Advanced/Treaps.mdx:107]: also comments explaining what's actually happening would be nice
or i'm just dumb

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: content/6_Advanced/Treaps.mdx (hunk 3 above)
Comment (dutinmeow) [reply content/6_Advanced/Treaps.mdx:106]: i may or may not have copied code with different variable names that I forgot to change

File: content/6_Advanced/Treaps.mdx [hunk 5]
```diff
@@ -92,34 +193,18 @@ int main() {
...
-
-- [Quora](https://tanujkhattar.wordpress.com/2016/01/10/treaps-one-tree-to-rule-em-all-part-1/)
-- [PPT](https://docs.google.com/presentation/d/14xgtdDWnIBwmJRAuIdZ8FvLZcX9uRxnNoGOGAQRDIvc/edit?usp=sharing)
-- Samuel Hsiang Guide (see resources)
-
-## Notes
-
-- For speed / memory, use arrays of fixed size rather than pointers.
-
 ## Problems
```
[the remark this answers is not in the document]
Comment (dutinmeow) [reply content/6_Advanced/Treaps.mdx:197]: What do you mean?

[the remark this answers is not in the document]
Comment (maggie-j-liu) [reply content/6_Advanced/Treaps.mdx:197]: use one of these 
image

**Comment by SansPapyrus683** [comment]
ok this is being nitpicky but do you think you can do away with the pointers

**Comment by dutinmeow** [comment]
but aren't treaps normally implemented with pointers?

**Comment by dutinmeow** [comment]
Even benq uses pointers in the example code

**Comment by SansPapyrus683** [comment]
wouldn't something easier to understand like a reference work?
idk i'm bad with low level stuff

**Comment by dutinmeow** [comment]
I think references would be even more confusing
seeing as I would have areference to a reference
I think the arrow notation for pointers is actually not too bad

File: content/6_Advanced/Treaps.mdx (hunk 5 above)
[the remark this answers is not in the document]
Comment (dutinmeow) [reply content/6_Advanced/Treaps.mdx:197]: I'll organize it in a future PR, along with other modules

**Comment by SansPapyrus683** [comment]
but hold on why do you need a pointer to a reference in the first place

**Comment by dutinmeow** [comment]
To modify the pointer

**Comment by SansPapyrus683** [comment]
ok then why don't you just change the `Node` struct to just have a reference to two other nodes, then have it take just direct references, and then modify those references?
idk how low-level memory stuff works so sorry if i sound dumb

**Comment by dutinmeow** [comment]
I don't think you would be able to freely dynamically allocate them. You would have to maintain a vector/deque of treap Nodes, which defeats the entire purpose of using pointers instead of indexes

**Comment by dutinmeow** [comment]
I can add comments to describe what pointers do and stuff if that makes it better?

**Comment by SansPapyrus683** [comment]
ok sure you can add comments but i still don't get it

> I don't think you would be able to freely dynamically allocate them

what's this mean

**Comment by dutinmeow** [comment]
When you insert a value, you create a new node. (Inserting would be something like `merge(root, new Node(val))`). To use references, you would have to store the actual nodes somewhere, such as in a vector, and then pass the reference in. [Click](https://stackoverflow.com/questions/3843373/binary-trees-in-c-using-references)

**Comment by dutinmeow** [comment]
Also see [here](https://www.quora.com/Why-do-most-C-implementations-of-binary-search-trees-use-pointers-to-store-nodes)

**Comment by SansPapyrus683** [comment]
image
uuuhhh i don't see anything

**Comment by dutinmeow** [comment]
😳 the link looked relavent so i put it there
quora is blocked on my computer

**Comment by SansPapyrus683** [comment]
alr
then you can just add some comments

**Comment by dutinmeow** [comment]
Basically you can't modify the object that a reference is referring to

**Review by SansPapyrus683 — APPROVED** [review]

**Comment by dutinmeow** [comment]
pls merge :pray:

File: content/6_Advanced/Treaps.mdx [hunk 6]
```diff
@@ -1,27 +1,138 @@
...
+	<Resource
+		source = "Algorithm Tutorial"
+		title = "Treaps : One Tree to Rule ’em all"
+		url = "https://tanujkhattar.wordpress.com/2016/01/10/treaps-one-tree-to-rule-em-all-part-1/"
+	>
+		Code and diagrams
+	</Resource>
+	<Resource
+		source = "Benq"
+		title = "Treap presentation"
```
Comment (bqi343) [inline content/6_Advanced/Treaps.mdx:36]: title doesn't correspond to URL

File: content/6_Advanced/Treaps.mdx [hunk 7]
```diff
@@ -1,27 +1,138 @@
...
+
 ## Splitting
 
+The `split` method takes in a pointer to the root of a treap $\texttt{root}$ and a value $x$, and returns two treaps denoted as $\texttt{left}$ and $\texttt{right}$. Like the name suggests, it splits the tree such that all nodes in $\texttt{left}$ have keys less than or equal to $x$ and all nodes in $\texttt{right}$ have keys greater than $x$. 
+
+We can now implement it recursively. Let the left child of a node $n$ be denoted as $n.\texttt{left}$ and the right child as $n.\texttt{right}$.
+- If $\texttt{root} \leq x$, then both the root and the left subtree belong to $\texttt{left}$. We now consider a call to `split` on $\texttt{root.right}$ and note its results as $\texttt{left'}$ and $\texttt{right'}$. Finally, $\texttt{left}$ contains $\texttt{left'}$ and $\texttt{right} = \texttt{right'}$
+- If $\texttt{root} > x$, then both the root and the right subtree belong to $\texttt{right}$. We now consider a call to `split` on $\texttt{root.left}$ and note its results as $\texttt{left'}$ and $\texttt{right'}$. Finally, $\texttt{right}$ contains $\texttt{right'}$ and $\texttt{left} = \texttt{left'}$
```
Comment (bqi343) [inline content/6_Advanced/Treaps.mdx:69]: Add periods at the end of sentences.

**Review by bqi343 — CHANGES_REQUESTED** [review]

**Review by SansPapyrus683 — APPROVED** [review]
this is pog

**Review by bqi343 — APPROVED** [review]

**Comment by dutinmeow** [comment]
yay it got merged :)

**Pull request merged**